clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,63 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = fmap snd . go
+ where
+  go :: HasCallStack => Term -> NormalizeSession (Bool, Term)
+  go e0 = inlineNonRepWorker e0 >>= \case
+    (True, e1) ->
+      return (True, e1)
+    (False, e1) ->
+      case e1 of
+        Var _ -> return (False, e1)
+        Data _ -> return (False, e1)
+        Literal _ -> return (False, e1)
+        Prim _ -> return (False, e1)
+        Lam v t -> second (Lam v) <$> go t
+        TyLam v t -> second (TyLam v) <$> go t
+        App a0 b0 -> do
+          (change0, a1) <- go a0
+          (change1, b1) <- go b0
+          return (change0 || change1, App a1 b1)
+        TyApp t typ -> second (`TyApp` typ) <$> go t
```

Comment:
I think applicative style might read nicer here, e.g.

```haskell
TyApp <$> go t <*> pure typ
```

Reply:
Although actually you only need the `second` because you carry around the `Bool`. So it should just become an `fmap`